dds/DCPS/RTPS/Spdp.cpp

```diff
@@ -3165,6 +3123,87 @@ Spdp::get_discovered_participant_ids(DCPS::RepoIdSet& results) const
   }
 }
 
+void
+Spdp::remove_lease_expiration_i(DiscoveredParticipantIter iter)
+{
+  for (std::pair<TimeQueue::iterator, TimeQueue::iterator> x = lease_expirations_.equal_range(iter->second.lease_expiration_);
+       x.first != x.second; ++x.first) {
+    if (x.first->second == iter->first) {
+      lease_expirations_.erase(x.first);
+      break;
+    }
+  }
+}
+
+void
+Spdp::update_lease_expiration_i(DiscoveredParticipantIter iter,
+                                const DCPS::MonotonicTimePoint& now)
+{
+  remove_lease_expiration_i(iter);
+
+  // Compute new expiration.
+  const DCPS::TimeDuration d =
+    rtps_duration_to_time_duration(iter->second.pdata_.leaseDuration,
+                                   iter->second.pdata_.participantProxy.protocolVersion,
+                                   iter->second.pdata_.participantProxy.vendorId);
+
+  iter->second.lease_expiration_ = now + d;
+
+  // Insert.
+  const bool cancel = !lease_expirations_.empty() && iter->second.lease_expiration_ < lease_expirations_.begin()->first;
+  const bool schedule = lease_expirations_.empty() || iter->second.lease_expiration_ < lease_expirations_.begin()->first;
+
+  lease_expirations_.insert(std::make_pair(iter->second.lease_expiration_, iter->first));
+
+  if (cancel) {
+    tport_->lease_expiration_processor_->cancel();
+  }
+  if (schedule) {
+    tport_->lease_expiration_processor_->schedule(d);
+  }
+}
+
+void
+Spdp::process_lease_expirations(const DCPS::MonotonicTimePoint& now)
+{
+  ACE_GUARD (ACE_Thread_Mutex, g, lock_);
+
+  for (TimeQueue::iterator pos = lease_expirations_.begin(), limit = lease_expirations_.end();
+       pos != limit && pos->first <= now;) {
+    DiscoveredParticipantIter part = participants_.find(pos->second);
+    // Pre-emptively erase so remove_discovered_participant will not modify lease_expirations_.
+    lease_expirations_.erase(pos++);
+
+    if (part == participants_.end()) {
+      continue;
+    }
+
+    if (DCPS::DCPS_debug_level) {
+      ACE_DEBUG((LM_WARNING,
+                 ACE_TEXT("(%P|%t) Spdp::process_lease_expirations() - ")
+                 ACE_TEXT("participant %C exceeded lease duration, removing\n"),
+                 DCPS::LogGuid(part->first).c_str()));
+    }
+
+#ifdef OPENDDS_SECURITY
+    ICE::Endpoint* sedp_endpoint = sedp_.get_ice_endpoint();
+    if (sedp_endpoint) {
+      stop_ice(sedp_endpoint, part->first, part->second.pdata_.participantProxy.availableBuiltinEndpoints);
+    }
+    ICE::Endpoint* spdp_endpoint = tport_->get_ice_endpoint();
+    if (spdp_endpoint) {
+      ICE::Agent::instance()->stop_ice(spdp_endpoint, guid_, part->first);
+    }
+    purge_handshake_deadlines(part);
+#endif
+    remove_discovered_participant(part);
+  }
+
+  if (!lease_expirations_.empty()) {
+    tport_->lease_expiration_processor_->schedule(lease_expirations_.begin()->first - now);
```

Comment:
Is now correct enough by this point in the function to use in calculating a relative wait period to the next expiration? We might want to update "now" as part of the loop or at least here at the end before the calculation. Carefully though, so we're not trying to schedule a negative relative wait time.